node/pkg/processor/reactor/manager.go

```diff
@@ -0,0 +1,228 @@
+	return &Manager[K]{
+		group:         group,
+		observationC:  observationC,
+		confirmationC: confirmationC,
+		setC:          setC,
+		reactors:      map[ethcommon.Hash]*ConsensusReactor[K]{},
+		gs:            gst.Get(),
+		gst:           gst,
+		config:        config,
+		handler:       handler,
+		storage:       storage,
+	}
+}
+
+func (p *Manager[K]) Run(ctx context.Context) error {
+	p.logger = supervisor.Logger(ctx)
+	p.logger = p.logger.With(zap.String("group", p.group))
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case gs := <-p.setC:
+			p.gsLock.Lock()
+			p.gs = gs
+
+			p.logger.Info("guardian set updated",
+				zap.Strings("set", p.gs.KeysAsHexStrings()),
+				zap.Uint32("index", p.gs.Index))
+			p.gst.Set(p.gs)
+			p.gsLock.Unlock()
+		case k := <-p.observationC:
+			messagesObservedTotal.WithLabelValues(p.group).Inc()
+			// Make sure to not reobserve messages that already have a persisted signed observation
+			if p.storage != nil {
+				existingObservation, _, found, err := p.storage.GetSignedObservation(k.MessageID())
+				if err != nil {
+					p.logger.Warn("failed to check db for existing signed observation", zap.String("message_id", k.MessageID()), zap.Stringer("digest", k.SigningMsg()), zap.Error(err))
+				}
+				if found {
+					p.logger.Debug("ignoring observation - already in storage", zap.String("message_id", k.MessageID()), zap.Stringer("digest_observed", k.SigningMsg()), zap.Stringer("digest_stored", existingObservation.SigningMsg()))
+					continue
+				}
+			}
+
+			digest := k.SigningMsg()
+			p.logger.Debug("received observation", zap.Stringer("digest", digest), zap.Any("observation", k))
+
+			var r *ConsensusReactor[K]
+			p.reactorsLock.Lock()
+			if reactor, exists := p.reactors[digest]; exists {
+				p.logger.Debug("sending observation to existing reactor", zap.Stringer("digest", digest))
+				r = reactor
+			} else {
+				p.logger.Debug("creating new reactor for observation", zap.Stringer("digest", digest))
+				p.gsLock.Lock()
+				r = NewReactor[K](p.group, p.config, p.gs, func(reactor *ConsensusReactor[K], oldState, newState State) {
+					p.transitionHook(digest, reactor, oldState, newState)
+				})
+				p.gsLock.Unlock()
+				err := supervisor.Run(ctx, fmt.Sprintf("reactor-%s", digest.String()), r.Run)
+				if err != nil {
+					p.logger.Error("failed to spawn reactor routine", zap.Error(err), zap.Stringer("digest", digest))
+				}
+				p.reactors[digest] = r
+				reactorNum.WithLabelValues(p.group).Set(float64(len(p.reactors)))
+			}
+			p.reactorsLock.Unlock()
+			r.ObservationChannel() <- k
+		case m := <-p.confirmationC:
+			digest := ethcommon.BytesToHash(m.Hash)
+
+			p.gsLock.Lock()
+			err := verifySignedObservation(p.group, m, p.gs)
+			p.gsLock.Unlock()
```

Comment:
I decided to keep the duplicate verification here because we don't want to create reactors for invalid signed observations (potential DoS).
I previously had a way to signal to the reactor that a signed observation was verified before but thought that increased the risk of someone using the code wrong.